csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -0,0 +1,506 @@
+        }
+
+        /// <summary>
+        /// Runs a train step on the loaded model for the given inputs.
+        /// </summary>
+        /// <param name="inputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the input values.</param>
+        /// <param name="outputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the output values.</param>
+        public void TrainStep(
+           IReadOnlyCollection<FixedBufferOnnxValue> inputValues,
+           IReadOnlyCollection<FixedBufferOnnxValue> outputValues)
+        {
+            TrainStep(_builtInRunOptions, inputValues, outputValues);
+        }
+
+        /// <summary>
+        /// Runs a train step on the loaded model for the given inputs. Uses the given RunOptions for this run.
+        /// </summary>
+        /// <param name="options">Specify <see cref="RunOptions"/> for step.</param>
+        /// <param name="inputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the input values.</param>
+        /// <param name="outputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the output values.</param>
+        public void TrainStep(
+            RunOptions options,
+            IReadOnlyCollection<FixedBufferOnnxValue> inputValues,
+            IReadOnlyCollection<FixedBufferOnnxValue> outputValues)
+        {
+            if (_trainOutputCount!= (ulong)outputValues.Count())
+            {
+                throw new ArgumentException($"Length of {nameof(outputValues)} ({outputValues.Count}) must match that of train model ({_trainOutputCount}).");
+            }
+            IntPtr[] inputValuesArray = GetOrtValuesHandles(inputValues, true);
+
+            IntPtr[] outputValuesArray = GetOrtValuesHandles(outputValues, false); /* pointers to Pre-allocated OrtValue instances */
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtTrainStep(_nativeHandle, options.Handle, (UIntPtr)inputValues.Count,
+                inputValuesArray, (UIntPtr)outputValues.Count, outputValuesArray));
+        }
+
+        /// <summary>
+        /// Runs the loaded model for the given inputs, and fetches the graph outputs.
+        /// </summary>
+        /// <param name="inputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the input values.</param>
+        /// <returns>Output Tensors in a Collection of NamedOnnxValue. User must dispose the output.</returns>
+        public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> TrainStep(
+            IReadOnlyCollection<FixedBufferOnnxValue> inputValues)
+        {
+            using (var cleanupList = new DisposableList<IDisposable>())
+            {
+                IntPtr[] inputValuesArray = GetOrtValuesHandles(inputValues, true);
+
+                var ortValues = new DisposableList<OrtValue>((int)_trainOutputCount);
```

Comment:
>ortValues [](http://example.com/codeflow?start=20&length=9)

Same here. We can add ortValues to the list above, can we not, otherwise the above list has a single entry.